sql/core/src/main/scala/org/apache/spark/sql/execution/StateImpl.scala

```diff
@@ -0,0 +1,70 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements.  See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License.  You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.spark.sql.execution
+
+import org.apache.spark.sql.State
+
+/** Internal implementation of the [[State]] interface */
+private[sql] class StateImpl[S](optionalValue: Option[S]) extends State[S] {
```

Comment:
I believe null can be a valid state value, so just taking null does not disambiguate whether state exists and is null v/s state does not exist. You would have to do `class StateImpl[S](value: S, isDefined: Boolean)`.  But this does not avoid object wrapping because, the Option wrapping is currently done by the StateStore implementation. We will have to change StateStore interfaces to avoid that. So I think its premature to do this in this PR.